### PR TITLE
mcp: add Extensions field to capabilities per SEP-2133

### LINKED ISSUE #777: Extension declaration for MCP Apps
The MCP Apps extension denotes that a client supporting MCP Apps should declare its capability as follows.

ref. https://github.com/modelcontextprotocol/ext-apps/blob/main/specification/2026-01-26/apps.mdx#clientserver-capability-negotiation

```json
{
  "method": "initialize",
  "params": {
    "protocolVersion": "2024-11-05",
    "capabilities": {
      "extensions": {
        "io.modelcontextprotocol/ui": {
          "mimeTypes": ["text/html;profile=mcp-app"]
        }
      }
    },
    "clientInfo": {
      "name": "claude-desktop",
      "version": "1.0.0"
    }
  }
}
```

However, [ClientCapabilities](https://pkg.go.dev/github.com/modelcontextprotocol/go-sdk@v1.2.0/mcp#ClientCapabilities) does not have a field for `extensions`, so we cannot implement a client that supports MCP Apps extension.

Similarly, MCP servers built with this SDK cannot verify client capabilities in `extensions`.

### PR DESCRIPTION
**Pull request by ymmt2005**
This implements [SEP-2133](https://modelcontextprotocol.io/community/seps/2133-extensions) and resolves #777.

## Summary of changes

- Add `Extensions` field to `ServerCapabilities` and `ClientCapabilities` structs per [SEP-2133](https://modelcontextprotocol.io/community/seps/2133-extensions)
- Add `AddExtension` helper methods on both types that normalize nil settings to empty maps.
  - [The spec requires an object, not null](https://modelcontextprotocol.io/community/seps/2133-extensions#negotiation:~:text=an%20empty%20object%20indicates%20no%20settings.)
- Include shallow-copy support via `maps.Clone` in the existing `clone()` methods
  - also include shallow-copy support for `Experimental` field.
- Add comprehensive tests: unit tests for nil-safety, capability preservation, and over-the-wire round-trip

## What's tested

- [x] `go test ./mcp/...` passes
- [x] Server conformance tests pass (30 scenarios, 40 assertions, 0 failures)
- [x] Client conformance tests match existing baseline (auth scenarios are known failures)

### CODE REVIEW

File: mcp/protocol.go [hunk 1]
```diff
@@ -213,8 +219,22 @@ type ClientCapabilities struct {
 }
 
 // clone returns a deep copy of the ClientCapabilities.
```
Comment (maciej-kisiel) [inline mcp/protocol.go]: Doc comment for clone was moved, please fix.

File: mcp/protocol.go [hunk 2]
```diff
@@ -213,8 +219,22 @@ type ClientCapabilities struct {
...
+	}
+	if settings == nil {
+		settings = map[string]any{}
+	}
+	c.Extensions[name] = settings
+}
+
 func (c *ClientCapabilities) clone() *ClientCapabilities {
 	cp := *c
+	cp.Extensions = maps.Clone(c.Extensions)
```
Comment (maciej-kisiel) [inline mcp/protocol.go:242]: According to `maps` documentation this is only a shallow copy and we need a deep one.

File: mcp/protocol.go [hunk 3]
```diff
@@ -1316,9 +1341,23 @@ type ServerCapabilities struct {
...
+	if settings == nil {
+		settings = map[string]any{}
+	}
+	c.Extensions[name] = settings
+}
+
 // clone returns a deep copy of the ServerCapabilities.
 func (c *ServerCapabilities) clone() *ServerCapabilities {
 	cp := *c
+	cp.Extensions = maps.Clone(c.Extensions)
```
Comment (maciej-kisiel) [inline mcp/protocol.go:1370]: Similarly here: please do a deep copy.

File: mcp/server_test.go [hunk 4]
```diff
@@ -455,6 +474,45 @@ func TestServerCapabilities(t *testing.T) {
 	}
 }
 
+func TestAddExtensionNilSafety(t *testing.T) {
```
Comment (maciej-kisiel) [inline mcp/server_test.go]: Isn't this sufficiently covered by the test cases added above?

File: mcp/server_test.go [hunk 5]
```diff
@@ -455,6 +474,45 @@ func TestServerCapabilities(t *testing.T) {
...
+		}
+		m, ok := settings.(map[string]any)
+		if !ok {
+			t.Fatalf("settings is %T, want map[string]any", settings)
+		}
+		if len(m) != 0 {
+			t.Fatalf("settings has %d entries, want 0", len(m))
+		}
+	})
+	t.Run("ClientCapabilities", func(t *testing.T) {
```
Comment (maciej-kisiel) [inline mcp/server_test.go]: This could be covered by extending `TestClientCapabilities` in `client_test.go` which is equivalent to the test that you modified for the server side above.

**Comment by maciej-kisiel** [comment]
Also, please fix lint.

File: mcp/protocol.go (hunk 3 above)
Comment (ymmt2005) [reply mcp/protocol.go:1370]: Thank you for the review.
It's not generally possible to deep copy `map[string]any`.

Shall we keep the shallow copy here?
Or shall we not copy anything for `Extensions`, just like `Experimental`?

File: mcp/server_test.go (hunk 4 above)
Comment (ymmt2005) [reply mcp/server_test.go]: yes, will remove this.

File: mcp/server_test.go (hunk 5 above)
Comment (ymmt2005) [reply mcp/server_test.go]: yes, will remove this.

File: mcp/protocol.go (hunk 1 above)
Comment (ymmt2005) [reply mcp/protocol.go]: addressed

File: mcp/protocol.go (hunk 3 above)
Comment (maciej-kisiel) [reply mcp/protocol.go:1370]: Right, maybe not impossible, but generally complex. Probably it's not warranted to incur this complexity at this point. Let's maybe adjust the doc comment of this function that the values in the `Extensions` and `Experimental` maps are shallow copied and adjust doc comments for `Extensions`, `Experimental` and `AddExtension` that the passed maps should not be modified after the call.

Same for `ClientCapabilities`.

Comment (ymmt2005) [reply mcp/protocol.go:1370]: Got it. Pushed a fix.

**Comment by ymmt2005** [comment]
Addressed all review comments, including the lint error.

**Review by maciej-kisiel — APPROVED** [review]

**Comment by maciej-kisiel** [comment]
Thanks for contributing!
